File: 03-sliding-window/02-longest-substring-without-repeating-characters/Solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char,int> lastSeen;
        int l = 0, best = 0;
        for (int r = 0; r < (int)s.size(); r++) {
            char c = s[r];
            if (lastSeen.count(c) && lastSeen[c] >= l) {
                l = lastSeen[c] + 1;
            }
            lastSeen[c] = r;
            best = max(best, r - l + 1);
        }
        return best;
    }
};

File: 03-sliding-window/02-longest-substring-without-repeating-characters/Solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        last_seen: dict[str, int] = {}
        l = 0
        best = 0
        for r, c in enumerate(s):
            if c in last_seen and last_seen[c] >= l:
                l = last_seen[c] + 1
            last_seen[c] = r
            best = max(best, r - l + 1)
        return best
